miasm2/core/objc.py

```diff
@@ -0,0 +1,1712 @@
+
+    @property
+    def void_ptr(self):
+        """Return the void* type"""
+        return self.get_type(('void*',))
+
+
+def ast_eval_size(ast, types_mngr):
+    """Evaluates the size of a C ast object
+
+    @ast: parsed pycparser.c_ast object
+    @types_mngr: types manager
+    """
+
+    if isinstance(ast, c_ast.TypeDecl):
+        result = ast_eval_size(ast.type, types_mngr)
+    elif isinstance(ast, c_ast.PtrDecl):
+        void_type = types_mngr.void_ptr
+        result = void_type.size
+    elif isinstance(ast, c_ast.IdentifierType):
+        obj = types_mngr.get_type(tuple(ast.names))
+        result = obj.size
+    else:
+        raise NotImplementedError('TODO')
+    return result
+
+
+def fix_recursive_objects(types_mngr, obj):
+    """Replace _ObjCRecurse objects by its parent"""
+
+    void_type = types_mngr.void_ptr
+
+    if isinstance(obj, ObjCStruct):
+        for i, (name, fieldtype, offset, size) in enumerate(obj.fields):
+            fieldtype = fix_recursive_objects(types_mngr, fieldtype)
+            obj.fields[i] = (name, fieldtype, offset, size)
+    elif isinstance(obj, ObjCDecl):
+        return obj
+    elif isinstance(obj, ObjCPtr):
+        target_obj = fix_recursive_objects(types_mngr, obj.objtype)
+        obj = ObjCPtr(obj.name, target_obj, void_type.align, void_type.size)
+    elif isinstance(obj, ObjCArray):
+        target_obj = fix_recursive_objects(types_mngr, obj.objtype)
+        obj = ObjCArray(target_obj, obj.elems)
+    elif isinstance(obj, ObjCUnion):
+        for i, (name, fieldtype, offset, size) in enumerate(obj.fields):
+            fieldtype = fix_recursive_objects(types_mngr, fieldtype)
+            obj.fields[i] = (name, fieldtype, offset, size)
+    elif isinstance(obj, _ObjCRecurse):
+        obj = types_mngr.get_type((obj.name,))
+    else:
+        raise NotImplementedError("Unknown type")
+    return obj
+
+
+def ast_eval_int(ast, types_mngr):
+    """Eval a C ast object integer
+
+    @ast: parsed pycparser.c_ast object
+    @types_mngr: types manager
+    """
+
+    if isinstance(ast, c_ast.BinaryOp):
+        left, right = ast_eval_int(ast.left, types_mngr), ast_eval_int(
+            ast.right, types_mngr)
+        if ast.op == '*':
+            result = left * right
+        elif ast.op == '/':
+            assert left % right == 0
+            result = left / right
+        elif ast.op == '+':
+            result = left + right
+        elif ast.op == '-':
+            result = left - right
+        else:
+            raise NotImplementedError("Not implemented!")
+    elif isinstance(ast, c_ast.UnaryOp):
+        if ast.op == 'sizeof' and isinstance(ast.expr, c_ast.Typename):
+            result = ast_eval_size(ast.expr.type, types_mngr)
+        else:
+            raise NotImplementedError("Not implemented!")
+
+    elif isinstance(ast, c_ast.Constant):
+        result = int(ast.value, 0)
+    elif isinstance(ast, c_ast.Cast):
+        # TODO: Can trunc intergers?
```

Comment:
For the TODO: wontfix:
I will see in future if we often face declaration with casting in integers of sizes for arrays.